File: src/gfx/src/macos/Palette.swift
import Math

public struct Palette {
  let bits: UInt32 = 32
  public var palette: [UInt32] = []
  public var count: Int
  public var data: [UInt32] = []

  init(single: UInt32, count: Int) {
    self.palette = [single]
    self.count = count
    self.data = []
  }

  init(uncompressed: [UInt32]) {
    count = uncompressed.count
    var seen = Set<UInt32>()
    for value in uncompressed {
      if !seen.contains(value) {
        seen.insert(value)
        palette.append(value)
      }
    }

    var bits = Int(ceil(x: log2(Float(palette.count))))
    var bitCursor = 0

    if bits == 0 {
      return  // No need to compress if there is only one value
    }

    var size = (uncompressed.count * bits) / Int(self.bits) + 1
    data = Array(repeating: 0, count: size)

    for value in uncompressed {
      let outerOffset = bitCursor / Int(self.bits)
      let innerOffset = bitCursor % Int(self.bits)

      if outerOffset >= size {
        break
      }

      // Find the index of value in palette
      let paletteIndex = UInt32(palette.firstIndex(of: value) ?? 0)
      // Create a proper mask with correct operator precedence
      let mask = (UInt32(1) << UInt32(bits)) - UInt32(1)

      let remainingBitsInCurrentUInt32 = Int(self.bits) - innerOffset

      if bits <= remainingBitsInCurrentUInt32 {
        data[outerOffset] |= (paletteIndex << innerOffset) & (mask << innerOffset)
      } else {
        let bitsInCurrent = remainingBitsInCurrentUInt32
        let bitsInNext = bits - bitsInCurrent

        // First part in current UInt32
        data[outerOffset] |= (paletteIndex << innerOffset)

        // Second part in next UInt32
        if outerOffset + 1 < size {
          data[outerOffset + 1] |= (paletteIndex >> bitsInCurrent) & ((1 << bitsInNext) - 1)
        } else {
          fatalError("out of bounds of data array")
        }
      }

      bitCursor += bits
    }
  }

  public func decompress() -> [UInt32] {
    if palette.isEmpty {
      return []  // Handle empty palette case to avoid log2 of 0
    }
    var bits = UInt32(ceil(x: log2(Float(palette.count))))
    var decompressedData: [UInt32] = []
    var bitCursor: UInt32 = 0

    let totalBitsInCompressedData = UInt32(data.count) * self.bits
    let maxValuesPossible = totalBitsInCompressedData / bits

    for _ in 0..<maxValuesPossible {
      let outerOffset = Int(bitCursor / self.bits)
      let innerOffset = Int(bitCursor % self.bits)

      var paletteIndex: UInt32 = 0

      let remainingBits = Int(self.bits) - innerOffset

      if bits <= remainingBits {
        if outerOffset < data.count {
          paletteIndex = (data[outerOffset] >> innerOffset) & ((1 << UInt32(bits)) - 1)
        } else {
          break  // Out of data bounds, stop decompression
        }
      } else {
        let bitsInCurrent = UInt32(remainingBits)
        let bitsInNext = bits - bitsInCurrent

        if outerOffset < data.count {
          paletteIndex |= (data[outerOffset] >> innerOffset) & ((1 << UInt32(bitsInCurrent)) - 1)
        } else {
          break  // Out of data bounds, stop decompression
        }
        if outerOffset + 1 < data.count {
          paletteIndex |= (data[outerOffset + 1] << UInt32(bitsInCurrent)) & ((1 << Int(bits)) - 1)
        } else {
          fatalError(
            "Warning: Palette index out of bounds: \(paletteIndex), palette count: \(palette.count)"
          )
        }
      }

      if UInt32(paletteIndex) < palette.count {
        decompressedData.append(palette[Int(paletteIndex)])
      } else {
        fatalError(
          "Warning: Palette index out of bounds: \(paletteIndex), palette count: \(palette.count)")
      }
      bitCursor += (bits)
    }
    return decompressedData
  }
  public subscript(index: UInt32) -> UInt32 {
    // 1. Input Validation
    precondition(index >= 0 && index < count, "Index \(index) out of bounds (count: \(count))")

    // 2. Handle Simple Cases (No Compressed Data)
    guard !palette.isEmpty else {
      // If count > 0 but palette is empty, this is an invalid state.
      fatalError("Palette is empty, cannot access index \(index).")
    }
    if data.isEmpty {
      // This means all values are the same (palette[0])
      return palette[0]
    }

    // 3. Calculate Bits Per Index (must match compression)
    // Use Swift.max(1, ...) to ensure bits is at least 1
    let bits = UInt32(Swift.max(1, UInt32(ceil(x: log2(Float(palette.count))))))
    guard bits > 0 else {
      // Should be caught by data.isEmpty check
      fatalError("UInt32ernal Error: bits calculated as 0 or less despite having data.")
    }

    // 4. Calculate Bit Position for the requested index
    let bitCursor = UInt32(index * bits)
    let outerOffset = Int(bitCursor / self.bits)
    let innerOffset = Int(bitCursor % self.bits)

    // 5. Safety Check: Ensure we don't read past the allocated data
    // Note: We might need outerOffset OR outerOffset+1
    let requiredDataIndex = (bitCursor + bits - 1) / self.bits  // The last data index needed
    precondition(
      requiredDataIndex < data.count,
      "Calculated data index \(requiredDataIndex) is out of bounds for data size \(data.count) when accessing index \(index)"
    )

    // 6. Extract the Palette Index from 'data' (Logic adapted from decompress)
    var paletteIndex: UInt32 = 0
    let remainingBits = UInt32(Int(self.bits) - innerOffset)
    let mask = (UInt32(1) << bits) - 1  // Mask to extract exactly 'bits' bits

    if bits <= remainingBits {
      // Value fits entirely in the current UInt32
      paletteIndex = (data[outerOffset] >> innerOffset) & mask
    } else {
      // Value spans across two UInt32s
      let bitsInCurrent = remainingBits
      let bitsInNext = bits - bitsInCurrent

      // Get lower bits from the current UInt32
      let part1 = data[outerOffset] >> innerOffset  // No mask needed here yet

      // Get upper bits from the next UInt32 (already checked bounds with requiredDataIndex)
      let maskNext = (UInt32(1) << bitsInNext) - 1
      let part2 = data[outerOffset + 1] & maskNext

      // Combine the parts
      paletteIndex = part1 | (part2 << bitsInCurrent)
      // Apply final mask for robustness (optional if logic is perfect, but safer)
      paletteIndex &= mask
    }

    // 7. Look Up Value in Palette
    let paletteArrayIndex = Int(paletteIndex)
    precondition(
      paletteArrayIndex >= 0 && paletteArrayIndex < palette.count,
      "Extracted palette index \(paletteIndex) is out of bounds for palette size \(palette.count) at index \(index)."
    )

    return palette[paletteArrayIndex]
  }
}

public struct PaletteId: Hashable {
  public var id: UInt32
  public init(id: UInt32) {
    self.id = id
  }
}

File: src/gfx/src/vk/palette.zig
const std = @import("std");

pub const Palette = struct {
    const Self = @This();

    bits: u32 = 32,
    palette: std.ArrayList(u32),
    count: usize,
    data: std.ArrayList(u32),
    allocator: *std.mem.Allocator,

    pub fn initSingle(allocator: *std.mem.Allocator, single: u32, count: usize) !Self {
        var palette = std.ArrayList(u32).init(allocator);
        try palette.append(single);

        return Self{
            .allocator = allocator,
            .palette = palette,
            .count = count,
            .data = std.ArrayList(u32).init(allocator),
        };
    }

    pub fn initUncompressed(allocator: *std.mem.Allocator, uncompressed: []const u32) !Self {
        var palette = std.ArrayList(u32).init(allocator);
        var seen = std.AutoHashMap(u32, void).init(allocator);
        defer seen.deinit();

        // Build palette of unique values
        for (uncompressed) |value| {
            if (!seen.contains(value)) {
                try seen.put(value, {});
                try palette.append(value);
            }
        }

        const bits_needed = if (palette.items.len <= 1) @as(u5, 0) else blk: {
            var bits: u5 = 0;
            var len = palette.items.len - 1;
            while (len > 0) : (len >>= 1) {
                bits += 1;
            }
            break :blk bits;
        };

        if (bits_needed == 0) {
            // No need to compress if there's only one value
            return Self{
                .allocator = allocator,
                .palette = palette,
                .count = uncompressed.len,
                .data = std.ArrayList(u32).init(allocator),
            };
        }

        const size = (uncompressed.len * bits_needed) / 32 + 1;
        var data = try std.ArrayList(u32).initCapacity(allocator, size);
        data.appendNTimesAssumeCapacity(0, size);

        var bit_cursor: usize = 0;

        for (uncompressed) |value| {
            const outer_offset = bit_cursor / 32;
            const inner_offset = bit_cursor % 32;

            if (outer_offset >= size) {
                break;
            }

            // Find index of value in palette
            const palette_index: u32 = blk: {
                var i: usize = 0;
                while (i < palette.items.len) : (i += 1) {
                    if (palette.items[i] == value) break :blk @as(u32, i);
                }
                break :blk 0; // Default fallback (should not happen)
            };

            const mask = (@as(u32, 1) << @as(u5, @intCast(bits_needed))) - 1;
            const remaining_bits = 32 - inner_offset;

            if (bits_needed <= remaining_bits) {
                const shift_amount1 = @as(u5, inner_offset);
                data.items[outer_offset] |= (palette_index << shift_amount1) & (mask << shift_amount1);
            } else {
                const bits_in_current = remaining_bits;
                const bits_in_next = bits_needed - bits_in_current;

                // First part in current u32
                const shift_amount2 = @as(u5, inner_offset);
                data.items[outer_offset] |= (palette_index << shift_amount2);

                // Second part in next u32
                if (outer_offset + 1 < size) {
                    const shift_amount3 = @as(u5, bits_in_current);
                    const shift_amount4 = @as(u5, bits_in_next);
                    data.items[outer_offset + 1] |= (palette_index >> shift_amount3) &
                        ((@as(u32, 1) << shift_amount4) - 1);
                } else {
                    @panic("out of bounds of data array");
                }
            }

            bit_cursor += bits_needed;
        }

        return Self{
            .allocator = allocator,
            .palette = palette,
            .count = uncompressed.len,
            .data = data,
        };
    }

    pub fn deinit(self: *Self) void {
        self.palette.deinit();
        self.data.deinit();
    }

    pub fn decompress(self: *const Self) ![]u32 {
        if (self.palette.items.len == 0) {
            return &[_]u32{};
        }

        const bits_needed = if (self.palette.items.len <= 1) @as(u5, 1) else blk: {
            var bits: u5 = 0;
            var len = self.palette.items.len - 1;
            while (len > 0) : (len >>= 1) {
                bits += 1;
            }
            break :blk bits;
        };

        var result = try self.allocator.alloc(u32, self.count);
        errdefer self.allocator.free(result);

        // If there's only one value, fill with it
        if (self.palette.items.len == 1) {
            @memset(result, self.palette.items[0]);
            return result;
        }

        // If no compressed data, return empty
        if (self.data.items.len == 0) {
            return result;
        }

        var bit_cursor: u32 = 0;
        const total_bits = @as(u32, self.data.items.len) * 32;
        const max_values = @min(total_bits / bits_needed, @as(u32, self.count));

        var i: usize = 0;
        while (i < max_values and i < result.len) : (i += 1) {
            const outer_offset = bit_cursor / 32;
            const inner_offset = bit_cursor % 32;

            var palette_index: u32 = 0;
            const remaining_bits = 32 - inner_offset;

            if (bits_needed <= remaining_bits) {
                if (outer_offset < self.data.items.len) {
                    const shift_amount1 = @as(u5, inner_offset);
                    palette_index = (self.data.items[outer_offset] >> shift_amount1) &
                        ((@as(u32, 1) << bits_needed) - 1);
                } else {
                    break; // Out of data bounds
                }
            } else {
                const bits_in_current = @as(u32, @intCast(remaining_bits));
                const bits_in_next = bits_needed - bits_in_current;

                if (outer_offset < self.data.items.len) {
                    const shift_amount2 = @as(u5, inner_offset);
                    const shift_amount3 = @as(u5, bits_in_current);
                    palette_index |= (self.data.items[outer_offset] >> shift_amount2) &
                        ((@as(u32, 1) << shift_amount3) - 1);
                } else {
                    break; // Out of data bounds
                }

                if (outer_offset + 1 < self.data.items.len) {
                    const shift_amount4 = @as(u5, bits_in_next);
                    const shift_amount5 = @as(u5, bits_in_current);
                    palette_index |= (self.data.items[outer_offset + 1] &
                        ((@as(u32, 1) << shift_amount4) - 1)) << shift_amount5;
                } else {
                    @panic("Palette index calculation error");
                }
            }

            if (palette_index < self.palette.items.len) {
                result[i] = self.palette.items[palette_index];
            } else {
                @panic("Palette index out of bounds");
            }

            bit_cursor += bits_needed;
        }

        return result;
    }

    pub fn get(self: *const Self, index: u32) u32 {
        std.debug.assert(index < self.count);

        // Handle simple cases
        if (self.palette.items.len == 0) {
            @panic("Palette is empty");
        }

        if (self.data.items.len == 0) {
            return self.palette.items[0];
        }

        // Calculate bits per index (must match compression)
        const bits_needed = std.math.max(1, blk: {
            var bits: u5 = 0;
            var len = self.palette.items.len - 1;
            while (len > 0) : (len >>= 1) {
                bits += 1;
            }
            break :blk bits;
        });

        const bit_cursor = index * bits_needed;
        const outer_offset = bit_cursor / 32;
        const inner_offset = bit_cursor % 32;

        // Safety check to ensure we don't read past allocated data
        const required_data_index = (bit_cursor + bits_needed - 1) / 32;
        std.debug.assert(required_data_index < self.data.items.len);

        // Extract the palette index from data
        var palette_index: u32 = 0;
        const remaining_bits = 32 - inner_offset;
        const mask = (@as(u32, 1) << bits_needed) - 1;

        if (bits_needed <= remaining_bits) {
            // Value fits entirely in the current u32
            const shift_amount1 = @as(u5, inner_offset);
            palette_index = (self.data.items[outer_offset] >> shift_amount1) & mask;
        } else {
            // Value spans across two u32s
            const bits_in_current = remaining_bits;
            const bits_in_next = bits_needed - bits_in_current;

            // Get lower bits from current u32
            const shift_amount2 = @as(u5, inner_offset);
            const part1 = self.data.items[outer_offset] >> shift_amount2;

            // Get upper bits from next u32
            const shift_amount3 = @as(u5, bits_in_next);
            const mask_next = (@as(u32, 1) << shift_amount3) - 1;
            const part2 = self.data.items[outer_offset + 1] & mask_next;

            // Combine parts
            const shift_amount4 = @as(u5, bits_in_current);
            palette_index = part1 | (part2 << shift_amount4);
            palette_index &= mask;
        }

        // Look up value in palette
        std.debug.assert(palette_index < self.palette.items.len);
        return self.palette.items[palette_index];
    }
};

pub const PaletteId = struct {
    id: u32,

    pub fn init(id: u32) PaletteId {
        return .{ .id = id };
    }

    pub fn hash(self: PaletteId) u64 {
        return @as(u64, self.id);
    }

    pub fn eql(self: PaletteId, other: PaletteId) bool {
        return self.id == other.id;
    }
};
